Parameterized unittests and benchmarks

burner opened the pull request:
Basically, Haskell Quickcheck with in build benchmarking for performance regression testing.

IMO phobos is missing something like Quickfix (randomized test data generation). Additionally, it would be nice to use the tests as a way to measure the performance of the functions and the code generation over time. 

The benchmark results are written to a file as csv. The program benchmarkplotter.d generates .dat and .gp files for all uniquely named benchmarks. gnuplot then generates a plot of the benchmark values over time. This could than be displayed on dlang.org to show our progress and be transparent about our performance.

make BUILD=benchmark enables all benchmarks. For more info see example in std/experimental/randomized_unittest_benchmark.d

![alt text](http://s22.postimg.org/h62nvhnv5/string_last_Index_Of_wstring_dstring.png)

``` d
/// The following examples show an overview of the given functionalities.
unittest
{
    void theFunctionToTest(int a, float b, string c)
    {
        // super expensive operation
        auto rslt = (a + b) * c.length;

        /* Pass the result to doNotOptimizeAway so the compiler
        can not remove the expensive operation, and thereby falsify the
        benchmark.
        */
        doNotOptimizeAway(rslt);

        debug
        {
            /* As the paramters to the function assume random values,
            $(D benchmark) allows to quickly test function with various input
            values. As the verification of computed value or state will at to
            the runtime of the function to benchmark, it makes sense to only
            execute these verifications in debug mode.
            */
            assert(c.length ? true : true);
        }
    }

    /* $(D benchmark) will run the function $(D theFunctionToTest) as often as
    possible in 1 second. The function will be called with randomly selected
    values for its parameters.
    */
    benchmark!theFunctionToTest();
}
```

dub: https://github.com/burner/std.benchmark

nordlow commented:
I've thought about implementing this. Good thing you've started it. Things I've been dreaming about:

- Benchmarker keeps track of total time spent and stops iteration (progression of test data sizes) after a caller-specified total time. To be exact this will require prediction of execution time for next data iteration (see next).
- Guess time-complexity by fitting samples to a list of models, say O(n), O(n^^2), ..., O(n*log(n)). This can be used in previous point to predict execution time given a specific data size.
- Multi-Dimensional Size Iterations?

burner commented:
* I'm not sure about the max run time stop part (but I accept PRs :-) )
* The time complexity stuff would be nice if we had a way to feed it back in into the source. (aka auto doc)
* what do you mean by multi-dim size iterators?

* I'm planning to have another unittest target running all the benchmark unitttest and have the result be convert into a gnuplot graph that we can show on the webpage.
* next I'm will create a unicode string generator

nordlow commented:
I update my previous comment.

burner commented:
anyone?

MartinNowak commented:
I have a very complete library for this kind of random testing, build after QuickCheck.
http://code.dlang.org/packages/qcheck
Andrei recently asked me to move the arbitrary part of it to std.random or so. Now you're aging something similar and duplicate existing work [std.benchmark](http://forum.dlang.org/thread/j5ojci$27ro$1@digitalmars.com).
While I think this is a good addition to phobos we should probably go the long route, integrate existing stuff with your code and phobos.
I could really take some help on getting arbitrary Phobos ready.

burner commented:
IMO the long road is wrong. qcheck, as far as I can see it, was not build for benchmarking. and std.benchmark was not build for testing. Both were never meant to work together, let alone in a consistent way. Therefore, I believe combining both will not fly and be a total mess. Making the combination work, would properly yield this PR. Just because these two projects exists does not mean they have to be used.

On the other hand PR shows a clear vision. It presents a clear track how to integrate all the way from phobos to marketing D and phobos on dlang.org. It works seamlessly with TypeTuple foreach loops. The Implementation and usage is trivial and was build to fit together. Extending the random value generation to new types is easy and flexible. benchmarkplotter.d is a nice showcase for D and especially phobos, IMO. It is properly trivial to integrate the benchmarking and record-keeping in the autotester (the BUILD target already exists). I can go on, how this PR will make all the bikeshedding about benchmarking superfluous and how it puts pressure one PRs to not worsen performance which leads to got marketing and so forth.

But most importantly, this PR is __ready now__. Not in a week or a year. __Now!__

p.s. Sorry for the rant, I'm sometimes just really annoyed by the scared, half-cocked decision making that seams to be governing D development. Instead of bold, risk taking, vision driven decision making. 

p.p.s. Even if this toolchain is a bust there is no risk. It is in internal and will not be exposed to the public, apart from the gnunplot graphs.

MartinNowak commented on std/traits.d:
```diff
@@ -6339,3 +6339,8 @@ unittest
+template RefType(T) 
+{
+	alias RefType = ref T;
+}
```
Adding a public symbol without documentation is a no-go.

burner replied:
moved that. as ref if strange anywhere

MartinNowak commented:
> qcheck, as far as I can see it, was not build for benchmarking. and std.benchmark was not build for testing

That's a pretty made up argument, benchmark was made for benchmarking and qcheck was made for generating random values. Now you want to benchmark with random value, well...
Anyhow, I was just trying to take the opportunity to push an import phobos addition (std.benchmark), and also wanted to take up the existing idea to move arbitrary to phobos.

MartinNowak commented:
I think you also mean [QuickCheck](http://www.cse.chalmers.se/~rjmh/QuickCheck) right?

MartinNowak commented on std/string.d:
```diff
@@ -1815,6 +1815,268 @@ ptrdiff_t lastIndexOfNeither(Char,Char2)(const(Char)[] haystack,
+            {
+                immutable theChar = cast(dchar)charsToSearch[toSearchFor];
+                auto idx = toSearchIn.indexOf(theChar);
+                ben.stop();
```
What's that `ben.stop()` doing here? Who is reactivating the stopwatch?

burner replied:
generator is reactivation the stopwatch

MartinNowak commented on std/string.d:
```diff
@@ -1815,6 +1815,268 @@ ptrdiff_t lastIndexOfNeither(Char,Char2)(const(Char)[] haystack,
+    immutable rounds = 20000;
+    auto rnd = Random(1);
+
+    foreach (S; TypeTuple!(string, wstring, dstring))
```
You know about the backend optimizer issue with long function bodies?
[Issue 7157 – Optimiser is O(n^2) w.r.t. function length](https://issues.dlang.org/show_bug.cgi?id=7157)
You better duplicate the foreach for each nested scope.

MartinNowak commented:
That's a lot of ugly boilerplate code for a single benchmark.

```d
    auto ben = Benchmark(format("string.lastIndexOf(%s,%s)", 
        S.stringof, R.stringof));

    auto generator = RndValueGen!(
        GenUnicodeString!(S, 10, 500), 
        GenUnicodeString!(R, 1, 10))
            (rnd, ben, rounds);

    foreach(S toSearchIn, R toSearchFor; generator)
    {
        auto idx = toSearchIn.lastIndexOf(toSearchFor);
        ben.stop();

        if (idx != -1) {
            assert(equal(
                toSearchIn[idx .. idx + to!S(toSearchFor).length],
                toSearchFor));
        }
    }
```
I'd suggest to redesign this around functions.
```d
void testee(S, R)(S toSearchIn, R toSearchFor)
{
    auto idx = toSearchIn.lastIndexOf(toSearchFor);
    if (idx != -1) {
        assert(equal(
            toSearchIn[idx .. idx + to!S(toSearchFor).length],
            toSearchFor));
    }
}

runBench!(testee, GenUnicodeString!(S, 10, 500), GenUnicodeString!(R, 1, 10))();
```

You could even infer the arguments from the function parameters (`runBench!(testee!(S, R))`) and go with a generic config for things like low and high, though runBench should still take generators.
https://github.com/MartinNowak/qcheck/blob/master/src/qcheck/config.d

MartinNowak commented:
If you're benchmarking functions like `indexOf` I'd be worried about the overhead of `clock_gettime` called by `StopWatch.stop/start()`.

MartinNowak commented on std/internal/test/randomized_unittest_benchmark.d:
```diff
@@ -0,0 +1,369 @@
+        the $(D theFunctionToTest). The member function $(D stop) of 
+        $(D Benchmark) stops the stopwatch. The stopwatch is automatically
+        resumed when the loop is continued. */
+        ben.stop(); 
```
If you're benchmarking, the that's a release build with assertions disabled.

MartinNowak commented:
At least you need to do something to handle noisy benchmark results, the simplest effective approach is to take the best out of 10 runs or so.
http://forum.dlang.org/post/jlsi14$1pi5$1@digitalmars.com
Please have a look at Andrei's benchmark module. It contains some very good concepts https://github.com/andralex/phobos/blob/benchmark/std/benchmark.d.

burner commented:
> I'd be worried about the overhead of clock_gettime

I do worry about this. I'm have not fund a good solution yet. maybe generate data for multiple runs and than just serve it.

the forum link didn't lead me anywhere

burner commented:
```d
runBench!(testee, GenUnicodeString!(S, 10, 500), GenUnicodeString!(R, 1, 10))();
```

I'm not to sure if that will reduce the code size in the long run. All that and testee needs to be include into a unittest and it is missing the name of the benchmark etc. 

I could collect some info with ctfe, but I rather like it obviouse. I will think about this some more.

burner commented:
thank you for reviewing

MartinNowak commented:
> you can't stop the stopwatch for the assertion.

No need to do that, assertion are disabled for release benchmarks anyhow.
That also solves part of the `clock_gettime` cost problem, if generating test data is cheap enough (or is pregenerated).

MartinNowak commented:
> missing the name of the benchmark

Just take the name of the function then.

MartinNowak commented:
> I'm not to sure if that will reduce the code size in the long run.

Leaving aside the amount of code, the coupling of the `RndValueGen` as `Range` with the `Benchmark` stopwatch is weird. Any benchmark framework I know is based on benchmarking functions, because that allows to encapsulate the logic for sampling and statistics in the benchmark framework.
And statistics are an often overlooked but important part of a useful benchmark, otherwise people draw optimization conclusions from noise.

burner commented:
in the meantime https://issues.dlang.org/show_bug.cgi?id=11632 broke my code.

```d
alias RefType = ref T;
```

does not work anymore. I may need to use a mixin to build the opApply delegate parameter list :(

schveiguy commented:
> does not work anymore

It shouldn't have ever worked. Use a pointer.

burner commented:
pointer are not an option opApply takes a delegate that has ref parameter

```d
int opApply(int delegate(ref int) dg);
```

schveiguy commented:
That's not what the unit test is doing that fails, though. It's creating a variable on the stack, and then creating a ref variable on the stack (which isn't supposed to be possible). For that, you need to use pointers, or a smart wrapper type.

If I read that referenced bug description, it appears that:

```D
alias ref int RefInt
```

compiled and essentially turned into a non-ref int. That it made a true ref for you in that unit test is clearly not what was intended.

I don't know what is supposed to happen with an alias of ref type is used in function parameter, or if that grammar even is supported. It appears you are depending on accidental features :) I think you will need to redesign that usage, unfortunately.

burner commented:
thanks for the description, I forgot the part about the stack ref type. Unfortunately it also doesn't work if I create a function fun(RefType!int) . The grammar supports it, but than the grammar support pretty much everythind ;-) ....

schveiguy commented:
> The grammar supports it

Yes, you are right! I didn't realize that, and I don't know how the hell this can be allowed :)

For example, that means something like:

```D
alias Foo = ref int;
Foo x;
```

Is valid. What does that even mean?

I'll also note that not all storage classes work:

```D
alias Foo = lazy int; //  Error: basic type expected, not lazy
```

burner commented:
D is fun sometimes :smile:

burner commented:
big update:
* use candlesticks min 0.25 quantil median 0.5 quantil max
* only display single call runtime

usage:
make -f posix.mak DMD=../dmd/src/dmd unittest BUILD=benchmark
dmd benchmarkplotter.d
./benchmarkplotter std/string.d_benchmark.csv

```bash
for i in *.gp; do
    gnuplot $i
done

yourImageViewer *.png
```

@andralex

DmitryOlshansky commented:
this is getting old. What about Andrei's std.benchmark does it bear any correlation with this work?

burner commented:
@DmitryOlshansky I know, but nobody seams to be interested in actually benchmarking there code, or whatever.

It has nothing to do with Andrei's std.benchmark.

DmitryOlshansky commented:
> @DmitryOlshansky I know, but nobody seams to be interested in actually benchmarking there code, or whatever.

I do. The thing is this pull is down to 3rd page of ~40 pulls stuck in perpetual limbo. I'm killing whatever stinks bad enough to pass away. Still I guess ~5% of reviewers will ever see this pull today. Open a NG discussion to reignite this?

burner commented:
will do

burner commented:
http://forum.dlang.org/post/vzcvwrbqpeamtnopmtoa@forum.dlang.org

andralex commented on benchmarkplotter.d:
```diff
@@ -0,0 +1,120 @@
+	}
+}
+
+auto gnuplot = " set title \"%s\"
```
use heredoc strings so you don't need to escape

andralex commented on benchmarkplotter.d:
```diff
@@ -0,0 +1,120 @@
+{
+	if (args.length == 1)
+	{
+		writeln("You need to pass in a a filename");
```
`stderr.`

andralex commented on benchmarkplotter.d:
```diff
@@ -0,0 +1,120 @@
+		return 1;
+	}
+
+	import std.csv;	
```
list symbols used

andralex commented on benchmarkplotter.d:
```diff
@@ -0,0 +1,120 @@
+
+	Entry[][string] entries;
+
+	foreach(entry; csvReader!Entry(cast(string)read(args[1])))
```
`readText` instead of casted read

andralex commented on std/internal/test/randomized_unittest_benchmark.d:
```diff
@@ -0,0 +1,495 @@
+    /* a, b and c will have random created inside $(D generator) that uses
+    $(D rnd) as source of randomness
+    */
+    foreach(int a, float b, string c; generator)     {
```
space after `foreach` throughout

andralex commented on std/internal/test/randomized_unittest_benchmark.d:
```diff
@@ -0,0 +1,495 @@
+import std.typetuple : TypeTuple;
+import std.traits;
+
+private auto avgStopWatchTime() {
```
you don't want the average, but the minimum (or the mode)

andralex commented on std/internal/test/randomized_unittest_benchmark.d:
```diff
@@ -0,0 +1,495 @@
+private TickDuration getQuantilTick(double q)(TickDuration[] ticks)
+    pure @safe
+{
+    size_t idx = cast(size_t)(ticks.length * q);
```
auto

burner replied:
why not size_t? I mean I cast it to size_t

andralex replied:
`size_t` is repeated

andralex commented on std/internal/test/randomized_unittest_benchmark.d:
```diff
@@ -0,0 +1,495 @@
+        dummy.stop();
+    }
+
+    return to!("nsecs",real)(sw.peek()) / numRounds;
```
space after comma throughout

andralex commented on std/internal/test/randomized_unittest_benchmark.d:
```diff
@@ -0,0 +1,495 @@
+    }
+    else
+    {
+        return (ticks[idx] + ticks[idx-1]) / 2;
```
space around operators: `idx - 1`

andralex commented on std/internal/test/randomized_unittest_benchmark.d:
```diff
@@ -0,0 +1,495 @@
+        TickDuration.from!"seconds"(3),
+        TickDuration.from!"seconds"(4),
+        TickDuration.from!"seconds"(5)
+    ];
```
`auto ticks = [1, 2, 3, 4, 5].map!(n => TickDuration.from!"seconds"(n)).array;`

andralex commented on std/internal/test/randomized_unittest_benchmark.d:
```diff
@@ -0,0 +1,495 @@
+    TickDuration[] ticks; // the stopped times, there will be rounds ticks
+    size_t ticksIndex = 0; // the index into ticks
+
+    alias stopWatch this;
```
why is the benchmark a subtype of its stopwatch?

burner replied:
to call start and stop on Benchmark

andralex replied:
use forwarding instead of subtyping

andralex commented on std/internal/test/randomized_unittest_benchmark.d:
```diff
@@ -0,0 +1,495 @@
+        if (this.ticksIndex < ticks.length)
+        {
+            this.ticks[this.ticksIndex] = this.stopWatch.peek();
+            ++this.ticksIndex;
```
why `this` everywhere?

burner replied:
because I like to know where my variables come from

andralex commented on std/internal/test/randomized_unittest_benchmark.d:
```diff
@@ -0,0 +1,495 @@
+    ~this()
+    {
+        import std.stdio : File;
+        if (!this.dontWrite)
```
`if (this.dontWrite) return;`

then do everything else

andralex commented on std/internal/test/randomized_unittest_benchmark.d:
```diff
@@ -0,0 +1,495 @@
+    }
+}
+
+struct GenNum(T,T low, T high)
```
public and undocumented?

andralex replied:
space after comma - particularly jarring here where there's both space and no space after comma

andralex commented on std/internal/test/randomized_unittest_benchmark.d:
```diff
@@ -0,0 +1,495 @@
+    alias opCall this;
+}
+
+struct GenString(T, size_t low, size_t high)
```
public and undocummented

andralex commented on std/internal/test/randomized_unittest_benchmark.d:
```diff
@@ -0,0 +1,495 @@
+    }
+
+    alias opCall this;
+}
```
This seems inefficient to me: generate full Unicode tables and then select elements from them. It seems an awful lot simpler and cheaper to just generate one correct random Unicode character at a time.

andralex commented on std/internal/test/randomized_unittest_benchmark.d:
```diff
@@ -0,0 +1,495 @@
+}
+
+
+private pure nothrow @safe string delBuilder(T...)()
```
a lil comment would help private artifacts too

andralex commented on std/string.d:
```diff
@@ -2114,6 +2114,274 @@ ptrdiff_t lastIndexOfNeither(Char,Char2)(const(Char)[] haystack,
+            }
+        }
+    }
+}
```
This seems to cut against the grain: massive boilerplate of which 90% is identical across different benchmarks, yet impossible to factor away; the actual benchmarked code is lost in the noise.

Ideally benchmark definitions should go together with the functions being benchmarked, in the same manners as unittests. Something like:

```d
auto fun(...)
{
    ... 
}

unittest 
{
   ...
}

auto benchmark_fun(...)
{
    ...
}
```

The challenge here is to have the benchmark framework use introspection and cooperation with the definition of `benchmark_fun` to automate most of the drudgery of selecting arguments, etc. Google's and benchmark framework has benchmark functions take an argument of type `benchmark::State` which includes a variety of things that the benchmark itself gets to use. It seems to me that's a useful model to follow. See https://github.com/google/benchmark. We need considerably less syntactic ugliness because of better introspection.

andralex commented:
The short version is this doesn't seem like something I want to use. 

Backend is fine but doesn't include things like "doNotOptimizeAway" which is essential. The front-end doesn't seem simple and easy to use, and it requires extensive boilerplate. I can't point to one or two issues to fix in the front end save for a complete rethinking.

burner commented:
I just saw the cppcon video of Chandlers benchmark talk. In it, the doNotOptimizeAway part jumped into my eye immediately. I will rethink and redo the frontend. And then I will run dfmt.

burner commented:
@andralex I rewrote the frontend, and added a very simple doNotOptimizeAway (compare the PR description). I also moved this into std.experimental. I create a dub package, and currently I'm trying to figure out what the best way is to add parameter bisect to find input that makes a tested function throw.

Any input?

JakobOvrum commented:
Is there any generated documentation for this?

burner commented:
the autotester of cybershadow is building the doc preview, or do you mean something else?

JakobOvrum commented:
Thanks, didn't know about that feature :)

jmdavis commented:
I haven't done more than glance at this, so I can't really comment on what it's up to, but please don't use `TickDuration`. The current plan is to deprecate it as soon as the few functions in Phobos that use it in their API have been deprecated. `MonoTime` and `Duration` are intended to replace all uses of `TickDuration`.

burner commented:
Currently, I'm mostly developing of the dub package https://github.com/burner/std.benchmark 
I have already removed all dependencies on TickDuration.

veelo commented on std/experimental/randomized_unittest_benchmark.d:
```diff
@@ -0,0 +1,878 @@
+/// Ditto
+unittest
+{
+    /* This function takes to $(D Gen) types as parameter. These $(D Gen)
```
takes two .. as parameters

burner replied:
done

veelo commented on std/experimental/randomized_unittest_benchmark.d:
```diff
@@ -0,0 +1,878 @@
+{
+    /* This function takes to $(D Gen) types as parameter. These $(D Gen)
+     types are implicitly convertiable to the type given as the first template
+    type parameter. The second and thrid template parameter give the upper and
```
third

burner replied:
done

veelo commented on std/experimental/randomized_unittest_benchmark.d:
```diff
@@ -0,0 +1,878 @@
+
+To gain appropriate test coverage and to test unexpected inputs, randomized
+unittest are a possible approach.
+Additionally, they lend itself for reproducible benchmarking and performance
```
they lend themselves

burner replied:
done

veelo commented on std/experimental/randomized_unittest_benchmark.d:
```diff
@@ -0,0 +1,878 @@
+    }
+}
+
+/* Return $(D true) is the passed $(D T) is a $(D Gen) struct.
```
if

burner replied:
done

veelo commented on std/experimental/randomized_unittest_benchmark.d:
```diff
@@ -0,0 +1,878 @@
+
+/* Return $(D true) is the passed $(D T) is a $(D Gen) struct.
+
+A $(D Gen!T) is something that implicitly converts to $(D T), has a methods
```
a method

burner replied:
done

veelo commented on std/experimental/randomized_unittest_benchmark.d:
```diff
@@ -0,0 +1,878 @@
+/* Return $(D true) is the passed $(D T) is a $(D Gen) struct.
+
+A $(D Gen!T) is something that implicitly converts to $(D T), has a methods
+called $(D gen) accepting a $(D ref Random) and has an $(D bisect) method
```
a $(D bisect)

burner replied:
done

veelo commented on std/experimental/randomized_unittest_benchmark.d:
```diff
@@ -0,0 +1,878 @@
+/** This type will generate a $(D Gen!T) for all passed $(D T...).
+Every call to $(D genValues) will call $(D gen) of all $(D Gen) structs
+present in $(D values). The member $(D values) can be passed to every
+functions accepting $(D T...).
```
every function / all functions

burner replied:
done

veelo commented on std/experimental/randomized_unittest_benchmark.d:
```diff
@@ -0,0 +1,878 @@
+
+/** A template that turns a $(D T) into a $(D Gen!T) unless $(D T) is
+already a $(D Gen) or no $(D Gen) for given $(D T) is avaiable.
+*/
```
available

burner replied:
done

veelo commented on std/experimental/randomized_unittest_benchmark.d:
```diff
@@ -0,0 +1,878 @@
+$(D maxRuntime). It will count how often $(D T) is run in the duration and
+how long each run took to complete.
+
+Unless compiled in release mode, statictis will be printed to $(D stderr).
```
statistics

burner replied:
done

veelo commented on std/experimental/randomized_unittest_benchmark.d:
```diff
@@ -0,0 +1,878 @@
+how long each run took to complete.
+
+Unless compiled in release mode, statictis will be printed to $(D stderr).
+If compiled in release mode the statictis are appended to a file called
```
statistics

burner replied:
done

veelo commented on std/experimental/randomized_unittest_benchmark.d:
```diff
@@ -0,0 +1,878 @@
+$(D name).
+
+Params:
+    opts = An $(D BenchmarkOptions) instance that encompasses all possible
```
A $(D BenchmarkOptions)

burner replied:
done

veelo commented on std/experimental/randomized_unittest_benchmark.d:
```diff
@@ -0,0 +1,878 @@
+
+Params:
+    opts = An $(D BenchmarkOptions) instance that encompasses all possible
+        parameter of benchmark.
```
parameters

burner replied:
done

veelo commented on std/experimental/randomized_unittest_benchmark.d:
```diff
@@ -0,0 +1,878 @@
+        parameter of benchmark.
+    name = The name of the benchmark. The name is also used as filename to
+        save the benchmark results.
+    maxRuntime = The maximul time the benchmark is executed. The last run will
```
maximum

burner replied:
done

veelo commented on std/experimental/randomized_unittest_benchmark.d:
```diff
@@ -0,0 +1,878 @@
+    name = The name of the benchmark. The name is also used as filename to
+        save the benchmark results.
+    maxRuntime = The maximul time the benchmark is executed. The last run will
+        not be interrupted
```
fullstop

burner replied:
done

veelo commented on std/experimental/randomized_unittest_benchmark.d:
```diff
@@ -0,0 +1,878 @@
+        not be interrupted
+    rndSeed = The seed to the random number generator used to populate the
+        parameter passed to the function to benchmark.
+    rounds = The maximum numbers of times the callable $(D T) is called.
```
number

burner replied:
gone

veelo commented on std/experimental/randomized_unittest_benchmark.d:
```diff
@@ -0,0 +1,878 @@
+    assert(newLogger.line.indexOf("'b'") != -1);
+}
+
+/** A functions that makes sure that the passed parameter are not optimized
```
parameters are / parameter is

burner replied:
done

veelo commented on std/experimental/randomized_unittest_benchmark.d:
```diff
@@ -0,0 +1,878 @@
+}
+
+/** A functions that makes sure that the passed parameter are not optimized
+away by the compiler. This function is required as optimizing compiler are
```
compilers

burner replied:
done

veelo commented on std/experimental/randomized_unittest_benchmark.d:
```diff
@@ -0,0 +1,878 @@
+able to figure out that a variable is not actually used, and therefore the
+computation of the value of the variable can be removed from code. As
+benchmarking functions sometimes include computing values that are not
+actually used, this functions allows use to force the compiler not to remove
```
function allows us

burner replied:
done, Thank you, for all the corrections

burner commented:
* rebased
* donotoptimzeaway updated
* dub package up to data v0.0.6

greenify commented:
What is the status of this? Seems like detecting performance regression is starting to get an important topic, see #3922

burner commented:
I'm waiting for more reviews and somebody willing to be the review manager

wilzbach commented:
> I'm waiting for more reviews and somebody willing to be the review manager

I am happy to count the feedbacks for you. Just post an announcement on the newsgroup ;-)
(I will add a link to a voting platform)

burner commented:
@wilzbach awesome. I will rebase to make it merge again

wilzbach commented on posix.mak:
```diff
@@ -156,6 +162,7 @@ STD_PACKAGES = std $(addprefix std/,\
   experimental/allocator/building_blocks experimental/logger \
   experimental/ndslice \
   net \
+  randomized_unittest_benchmark \
```
experimental?

wilzbach commented on std/stringbenchmarks.d:
```diff
@@ -0,0 +1,345 @@
+module std.stringbenchmarks;
```
needs to go to experimental too :/

wilzbach commented on benchmarkplotter.d:
```diff
@@ -0,0 +1,120 @@
+import std.stdio;
```
maybe we can create a folder `tools`?

wilzbach replied:
needs shebang (for lazy people with rdmd)
```
#!/usr/bin/env rdmd
```

and maybe a short comment header explaining the program and author ;-)

wilzbach commented on etc/c/donotoptimizeaway/win64.mak:
```diff
@@ -0,0 +1,28 @@
+MODEL=64
+VCDIR=\Program Files (x86)\Microsoft Visual Studio 10.0\VC
```
will this work for all windows users?

wilzbach commented on std/string.d:
```diff
@@ -2359,6 +2359,7 @@ auto capitalize(S)(auto ref S s)
     });
 }
 
+
```
?

wilzbach commented on etc/c/donotoptimizeaway/donotoptimizeaway.c:
```diff
@@ -0,0 +1,29 @@
+#include <stdio.h>
```
Is there no way we can do this in D?
Currently there is a huge movement to get rid of `etc/c` ...

wilzbach commented on std/experimental/randomized_unittest_benchmark.d:
```diff
@@ -0,0 +1,877 @@
+{
+    import std.typetuple : TypeTuple;
+
+    import std.meta : aliasSeqOf; //TODO uncomment with next release
```
what's the story here?

wilzbach commented:
> @wilzbach awesome. I will rebase to make it merge again

I am happy to push people!
Feel free to make a newsgroup announcement - I will collect the feedback then ;-)
Looking at the [last announcement](http://forum.dlang.org/post/stesmdubyyldrseepqqz@forum.dlang.org) there are probably a few things to bear in mind
- title should have [phobos-experimental] tag
- short & concise description
- people want to preview an addition with dub

burner commented:
@wilzbach still feel like being review manager?

wilzbach commented:
> @wilzbach still feel like being review manager?

Sure - as said I am more than happy to ensure that feedback will be collected ;-)
Should I make the NG announcement or do you want to describe your project yourself?

burner commented:
If you could please start the NG thread. That would be awesome.

wilzbach commented:
> If you could please start the NG thread. That would be awesome.

I hope [this announcement](http://forum.dlang.org/post/zfqqowupwtabmjdjqjon@forum.dlang.org) describes your PR correctly.

JackStouffer commented:
There's something wrong with your docs, see the third example: http://dtest.thecybershadow.net/artifact/website-73c6475e4be442350a727d80e562bb2328f494b1-af1863fd56f736d1f27cee0862282c44/web/library-prerelease/std/experimental/randomized_unittest_benchmark.html

nomad-software commented:
IMHO the namespace needs a little work. `std.experimental.randomized_unittest_benchmark` is not very scalable and ruins the namespace for future work. I would suggest `std.experimental.unititest.benchmark.randomised` or something near. Then other work could be added around those specific modules.

JackStouffer commented:
You should remove all global selective imports. While DMD now warns about them, LDC users who import the library manually will still experience 314 with no warning.

JackStouffer commented on std/experimental/randomized_unittest_benchmark.d:
```diff
@@ -0,0 +1,886 @@
+    else static if (is(T : GenASCIIString!(S), S...))
+        alias ParameterToGen = T;
+    else
+        static assert(false);
```
Need an error message here

burner replied:
I will fix that

andralex commented:
Plopping this here to make sure: http://forum.dlang.org/post/npk8rb$13pk$1@digitalmars.com

burner commented:
@andralex IMO the current approach is superior to your suggests. 

1. You can access the surrounding non global scope to access additional no random data.
2. This approach also allows me create a patch makes it possible to seamlessly pass additional data as function arguments.
3. My approach does not require a different test runner. Your approach requires to test all functions in all modules if they are benchmarks. Not hard but unnecessarily complex, especially when we have the unit test runner.
4. Aggregate.median already exists. The additional tool I created even allows you to track the performance over time (see Figure at the top). 
5. If you can have functions you can have template functions. That allows you to use all their magic. Sure you can do that with your approach as well, but that would require to create another function that calls the benchmark function. And I would argue using unittests as that function is more consistent.
6. The binding between parameter value range and parameter is clearer in my approach. (Is Aggregate.median the value of float b?)
7. IMO your example is overusing UDAs. Sure UDAs are awesome, but that does not mean we have to invent yet another DSL.

andralex commented:
>@andralex IMO the current approach is superior to your suggests. 

A productive thing to do would be to look at other benchmark frameworks and see how yours stacks against them. But at the end I agree it may end up as you and I having different design sensibilities.

>1. You can access the surrounding non global scope to access additional no random data.

What I wrote was but a sketch, shouldn't be taken as a complete design. Clearly more features may be added (such as fixtures, before/after code and state, etc).

>2. This approach also allows me create a patch makes it possible to seamlessly pass additional data as function arguments.

The same can be trivially achieved with the benchmark you just fix `Param!3(3.14)`. What am I missing?

>3. My approach does not require a different test runner. Your approach requires to test all functions in all modules if they are benchmarks. Not hard but unnecessarily complex, especially when we have the unit test runner.

I agree that there's more work involved with introspecting the attributes. But that work will be reused across all benchmarks, making it easy to write many individual benchmarks. It's scalability and nicely applied.

>4. Aggregate.median already exists. The additional tool I created even allows you to track the performance over time (see Figure at the top).

Sure, but that's not a competitive trait; I assume such a feature would be present in any framework along with a few others (min, max, average, p90 come to mind).
 
>5. If you can have functions you can have template functions. That allows you to use all their magic. Sure you can do that with your approach as well, but that would require to create another function that calls the benchmark function. And I would argue using unittests as that function is more consistent.

I'm not sure how you mean that. Surely attributes can be applied to template functions?

>6. The binding between parameter value range and parameter is clearer in my approach. (Is Aggregate.median the value of float b?)

It is clearer because the function _is_ a benchmark, not a function _being_ benchmarked.

>7. IMO your example is overusing UDAs. Sure UDAs are awesome, but that does not mean we have to invent yet another DSL.

Now this is just a fallacy: https://en.wikipedia.org/wiki/Begging_the_question

andralex commented:
The main thing here is that an attribute-based framework benchmarks _existing_ functions that are not written necessarily to implement benchmark, whereas this PR defines a framework in which the user _must_ define benchmarks. It follows that an attribute-based framework needs less code from the user.

* This PR: "You write the benchmarks and I call them"
* Attribute-based: "You write your usual application code and mention what values they should be benchmarked with, and I'll take care of the rest".

There is no contest here.

burner commented:
> The main thing here is that an attribute-based framework benchmarks existing functions that are not 
> written necessarily to implement benchmark, whereas this PR defines a framework in which the user 
> must define benchmarks. It follows that an attribute-based framework needs less code from the user.

I think I do not understand what you mean. I mean

```D
int someFunctionYouWantToBenchmark(int a, int b);

unittest {
    void fun(int a, int b) {
        int r = someFunctionYouWantToBenchmark(a, b);
       ....
    }

    benchmark!fun();
}
```

At look at the list at stackoverflow and more closely at those that looked good. This PR has all there features (some are implicitly given by D itself), additionally I couldn't find any library that allows for user defined types. This PR allows that.

> Param!3(3.14) What am I missing?
You have two places to keep in sync, which is worse than one place. Being DRY and all. And 3 is not really clear IMO ;-)

> I agree that there's more work involved with introspecting the attributes.
That is not really a problem, if it were no library ever would be created. IMO UDA are harder to use in the long run (I have no data, just my opinion on UDAs from what I have seen and used in D libraries.)

> It is clearer because the function is a benchmark, not a function being benchmarked. 
I do not understand what you mean.

> There is no contest here.
Well sort of, not between us, but between our ideas. Because, I hope you would agree that providing both approaches is not a good idea, that means we either have to choose one, or invent something better.

burner commented:
superseeded by https://github.com/burner/benchmark
